test/coverage/utils/CallbackConnectionTest.cpp

```diff
@@ -23,16 +28,469 @@ class TestFixture : public testing::Test {
-	~TestFixture() = default;
+	CallbackTest() = default;
+	~CallbackTest() = default;
 
 	// Run once per execution of the test application.
 	// Used only for global setup outside of tests.
 	static void SetUpTestSuite() {}
 	static void TearDownTestSuite() {}
 };
 
-// TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(CallbackTest, EstablishDoesNotThrow) {
+	using namespace uprotocol::utils;
+
+	EXPECT_NO_THROW(callbacks::Connection<void>::establish([]() {}));
+}
+
+TEST_F(CallbackTest, EstablishLinkedPair) {
+	using namespace uprotocol::utils;
+
+	int call_count{0};
+
+	auto [handle, callable] = callbacks::Connection<void>::establish(
+	    [&call_count]() { ++call_count; });
+
+	EXPECT_TRUE(handle);
+	EXPECT_TRUE(callable);
+	EXPECT_EQ(call_count, 0);
+
+	EXPECT_NO_THROW(callable());
+
+	EXPECT_TRUE(handle);
+	EXPECT_TRUE(callable);
+	EXPECT_EQ(call_count, 1);
+
+	callable();
+	callable();
+
+	EXPECT_TRUE(handle);
+	EXPECT_TRUE(callable);
+	EXPECT_EQ(call_count, 3);
+}
+
+TEST_F(CallbackTest, DroppedHandlesBreakConnection) {
+	int call_count{0};
+
+	auto getPair = ([&call_count]() {
+		using namespace uprotocol::utils;
+
+		auto connected_pair = callbacks::Connection<void>::establish(
+		    [&call_count]() { ++call_count; });
+
+		// Handle
+		EXPECT_TRUE(std::get<0>(connected_pair));
+		// Callable
+		EXPECT_TRUE(std::get<1>(connected_pair));
+
+		return connected_pair;
+	});
+
+	// Drop the handle end via reset()
+	{
+		auto [handle, callable] = getPair();
+		handle.reset();
+		EXPECT_FALSE(handle);
+		EXPECT_FALSE(callable);
+		callable();
+		EXPECT_EQ(call_count, 0);
+	}
+
+	// Drop the callable end via reset()
+	{
+		auto [handle, callable] = getPair();
+		callable.reset();
+		EXPECT_FALSE(handle);
+		EXPECT_FALSE(callable);
+	}
+
+	// Drop the handle end by letting it go out of scope
+	{
+		using namespace uprotocol::utils;
+		callbacks::Connection<void>::Callable callable_outside;
+		static_assert(std::is_move_assignable_v<decltype(callable_outside)>);
+		EXPECT_FALSE(callable_outside);
+		{
+			auto [handle, callable] = getPair();
+			callable_outside = std::move(callable);
+			EXPECT_FALSE(callable);
+			EXPECT_TRUE(callable_outside);
+			// Handle dropped here
+		}
+		EXPECT_FALSE(callable_outside);
+		EXPECT_NO_THROW(callable_outside());
+		EXPECT_EQ(call_count, 0);
+	}
+
+	// Drop the callable end by letting it go out of scope
+	{
+		using namespace uprotocol::utils;
+		callbacks::Connection<void>::Handle handle_outside;
+		static_assert(std::is_move_assignable_v<decltype(handle_outside)>);
+		EXPECT_FALSE(handle_outside);
+		{
+			auto [handle, callable] = getPair();
+			handle_outside = std::move(handle);
+			EXPECT_FALSE(handle);
+			EXPECT_TRUE(handle_outside);
+			// Callable dropped here
+		}
+		EXPECT_FALSE(handle_outside);
+	}
+
+	// If multiple copies of the callable end are held, dropping one doesn't
+	// break the connection.
+	{
+		auto [handle, callable] = getPair();
+		auto other_callable = callable;
```

Comment:
Point of note, test code does not need to be Autosar compliant.

Reply:
Per Autosar A7-1-5:

> The auto specifier shall not be used apart from following cases: (1) to declare
that a variable has the same type as return type of a function call...

This use of `auto` is Autosar compliant.

Reply:
(I know I submitted another comment here, but it's not showing up so I'm trying again)

I must have needed more sleep when I wrote that earlier comment. The letter of the requirement in A7-1-5 would say that `auto other_callable = callable` is not allowed. However, the _meaning_ of exception 1 effectively applies here. There is no intervening code between the line before where a value is assigned auto from a function return and where that value is immediately copied to another value. The type inferences operate the same in both cases. There would be little to no value in the first use of `auto` it the second use were not allowed.

Reply:
Tracking request in #154